### PR TITLE
Add option to use netcdf global_berror files

### LINKED ISSUE #808: Extend GSI to read netcdf global_berror files
JEDI can read both binary and netcdf format global_berror files.  This issue is opened to document that changes needed to allow the GSI to read netcdf format global_berror files.

### PR DESCRIPTION
**Pull request by RussTreadon-NOAA**
**Description**
This PR adds the option for `gsi.x` to read netcdf format `global_berror` files.

Resolves #808 

Depends on 
 - [x] GSI-fix [#25](https://github.com/NOAA-EMC/GSI-fix/pull/25)
 - [x] g-w issue [#3128](https://github.com/NOAA-EMC/global-workflow/issues/3128)

**Type of change**

- [x] New feature (non-breaking change which adds functionality)

**How Has This Been Tested?**

ctests run on [WCOSS2 (Dogwood)](https://github.com/NOAA-EMC/GSI/issues/808#issuecomment-2499541430) with _Passed_ result for all tests.

NetCDF format `global_berror` files tested for various resolutions on Dogwood.  Runs using netcdf and binary `global_berror` files yield bitwise identical analysis increment files.

  
**Checklist**

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] New and existing tests pass with my changes

### CODE REVIEW

**Comment by RussTreadon-NOAA** [comment]
This PR will changed to `Ready for review` once 
- [x] GSI-fix PR [#25](https://github.com/NOAA-EMC/GSI-fix/pull/25) is merged into `develop`
- [x] g-w issue [#3128](https://github.com/NOAA-EMC/global-workflow/issues/3128) is closed
- [x] the `fix` submodule hash in `RussTreadon-NOAA:feature/netcdf_berror` is updated
- [x] `GSI_BINARY_SOURCE_DIR` is updated to point at the new directory for staged non-ASCII GSI fix files

File: src/gsi/m_berror_stats.f90
Comment (CoryMartin-NOAA) [inline src/gsi/m_berror_stats.f90]: can we remove this commented line?

File: src/gsi/m_berror_stats.f90
Comment (CoryMartin-NOAA) [inline src/gsi/m_berror_stats.f90]: do we need these two lines?

File: src/gsi/m_berror_stats.f90
Comment (CoryMartin-NOAA) [inline src/gsi/m_berror_stats.f90]: if we get this from mpimod, we don't need it defined here

File: src/gsi/m_berror_stats.f90
Comment (CoryMartin-NOAA) [inline src/gsi/m_berror_stats.f90]: do we need these lines?

File: src/gsi/m_berror_stats.f90
Comment (CoryMartin-NOAA) [inline src/gsi/m_berror_stats.f90]: remove the varq and varcw_out lines?

File: src/gsi/m_berror_stats.f90
Comment (CoryMartin-NOAA) [inline src/gsi/m_berror_stats.f90:736]: is it worth adding a warning print / failure if a cvars2d is defined but no entry in the file is expected? or is it fine that this is hard coded for just these 2?

File: src/gsi/m_berror_stats.f90
Comment (CoryMartin-NOAA) [inline src/gsi/m_berror_stats.f90]: remove commented out lines?

File: src/gsi/m_berror_stats.f90
Comment (CoryMartin-NOAA) [inline src/gsi/m_berror_stats.f90]: why float and not real + rkind? I assume r_kind is float?

File: src/gsi/m_berror_stats.f90
Comment (CoryMartin-NOAA) [inline src/gsi/m_berror_stats.f90]: if it's only for PE0, do we need `mype = ', mype`?

File: src/gsi/m_berror_stats.f90
Comment (CoryMartin-NOAA) [inline src/gsi/m_berror_stats.f90]: same question as above

File: src/gsi/m_berror_stats.f90
Comment (CoryMartin-NOAA) [inline src/gsi/m_berror_stats.f90:1147]: do we bother with including the chem stuff from Brad Weir in this PR?

File: src/gsi/m_nc_berror.f90
Comment (CoryMartin-NOAA) [inline src/gsi/m_nc_berror.f90]: no need for these global attributes?

File: src/gsi/m_nc_berror.f90
Comment (CoryMartin-NOAA) [inline src/gsi/m_nc_berror.f90]: do we need this write routine? is it exercised anywhere?

File: src/gsi/m_berror_stats.f90
Comment (RussTreadon-NOAA) [reply src/gsi/m_berror_stats.f90]: Commented out line removed at 68d52ac

File: src/gsi/m_berror_stats.f90
Comment (RussTreadon-NOAA) [reply src/gsi/m_berror_stats.f90]: Remove commented out lines at 68d52ac

File: src/gsi/m_berror_stats.f90
Comment (RussTreadon-NOAA) [reply src/gsi/m_berror_stats.f90]: Remove commented out line at 68d52ac

File: src/gsi/m_berror_stats.f90
Comment (RussTreadon-NOAA) [reply src/gsi/m_berror_stats.f90]: Remove commented out lines at 68d52ac

File: src/gsi/m_berror_stats.f90
Comment (RussTreadon-NOAA) [reply src/gsi/m_berror_stats.f90]: Remove commented out line at 68d52ac

File: src/gsi/m_berror_stats.f90
Comment (RussTreadon-NOAA) [reply src/gsi/m_berror_stats.f90]: Revert back to `develop` to use `real` and `r_kind`.  Make the same change on line 865.  Done at 68d52ac.

File: src/gsi/m_berror_stats.f90
Comment (RussTreadon-NOAA) [reply src/gsi/m_berror_stats.f90]: Revert back to `develop` to use `real` and `r_kind`.  Make the same change on line 865.  Done at 68d52ac.

File: src/gsi/m_berror_stats.f90
Comment (RussTreadon-NOAA) [reply src/gsi/m_berror_stats.f90]: Three write statements in `sethwlloz_` have the same structure - write limited to mpi rank 0 and prints mpi rank 0 to stdout.  By including the mpi rank in printout the user know the task from which the printout originated without needing to refer to the code.  Is this necessary?  No.  One can turn on task tracing via environment variables or command line arguments.  Remove `mype` from this and similar lines in `m_berror_stats.f90`.  Done at 68d52ac.

File: src/gsi/m_berror_stats.f90
Comment (RussTreadon-NOAA) [reply src/gsi/m_berror_stats.f90:1147]: `setcorchem_` is in `develop`.  That's not an argument for retaining it but just a mention that this PR did not add `setcorchem_`.

File: src/gsi/m_nc_berror.f90
Comment (RussTreadon-NOAA) [reply src/gsi/m_nc_berror.f90]: Remove commented lines.  Done at 68d52ac

File: src/gsi/m_nc_berror.f90
Comment (RussTreadon-NOAA) [reply src/gsi/m_nc_berror.f90]: Good catch.  Neither subroutine `write_berror_` nor its interface `nc_berror_write` are used in the GSI code.  Both removed at 68d52ac

File: src/gsi/m_berror_stats.f90
Comment (RussTreadon-NOAA) [reply src/gsi/m_berror_stats.f90:736]: Good point.  Only 2 of the 9 `call nc_berro_getpointer` have logic handling the `ier /= 0` case.  When `ier /= 0` is encountered for these 2 cases, program execution is terminated.  Error checking added for the 7 other  cases.  Done at 849b3b19

File: src/gsi/m_berror_stats.f90
Comment (RussTreadon-NOAA) [reply src/gsi/m_berror_stats.f90]: Remove commented out lines at 849b3b19

**Review by CoryMartin-NOAA — DISMISSED** [review]

**Comment by RussTreadon-NOAA** [comment]
Install `RussTreadon-NOAA:feature/netcdf_berror` at 0e07d71f and `develop` at 9ada88e8 on Dogwood, Hera, Hercules, and Orion.  Run ctests with the following results

_Dogwood (WCOSS2)_
```
Test project /lfs/h2/emc/da/noscrub/russ.treadon/git/gsi/berror/build
    Start 1: global_4denvar
    Start 2: rtma
    Start 3: rrfs_3denvar_rdasens
    Start 4: hafs_4denvar_glbens
    Start 5: hafs_3denvar_hybens
    Start 6: global_enkf
1/6 Test #3: rrfs_3denvar_rdasens .............   Passed  1207.91 sec
2/6 Test #6: global_enkf ......................   Passed  1331.62 sec
3/6 Test #5: hafs_3denvar_hybens ..............   Passed  1633.44 sec
4/6 Test #4: hafs_4denvar_glbens ..............   Passed  1813.80 sec
5/6 Test #1: global_4denvar ...................   Passed  2523.63 sec
6/6 Test #2: rtma .............................   Passed  3010.41 sec

100% tests passed, 0 tests failed out of 6

Total Test time (real) = 3010.53 sec
```

_Orion_
```
Test project /work2/noaa/da/rtreadon/git/gsi/pr810/build
    Start 1: global_4denvar
    Start 2: rtma
    Start 3: rrfs_3denvar_rdasens
    Start 4: hafs_4denvar_glbens
    Start 5: hafs_3denvar_hybens
    Start 6: global_enkf
1/6 Test #3: rrfs_3denvar_rdasens .............   Passed  967.15 sec
2/6 Test #6: global_enkf ......................   Passed  1207.74 sec
3/6 Test #2: rtma .............................   Passed  1867.69 sec
4/6 Test #5: hafs_3denvar_hybens ..............   Passed  3085.11 sec
5/6 Test #4: hafs_4denvar_glbens ..............   Passed  3267.06 sec
6/6 Test #1: global_4denvar ...................***Failed  4202.53 sec

83% tests passed, 1 tests failed out of 6

Total Test time (real) = 4202.56 sec
```
The `global_4denvar` test failed due to
```
The runtime for global_4denvar_hiproc_updat is 881.487325 seconds.  This has exceeded maximum allowable threshold time of 815.596658 seconds, resulting in Failure of timethresh2 the regression test.
```
`gsi.x` wall times are known to be highly variable on Orion, especially when running in the `/work/noaa/stmp` fileset.  This is not a fatal fail.


_Hera_
```
Test project /scratch1/NCEPDEV/da/Russ.Treadon/git/gsi/pr810/build
    Start 1: global_4denvar
    Start 2: rtma
    Start 3: rrfs_3denvar_rdasens
    Start 4: hafs_4denvar_glbens
    Start 5: hafs_3denvar_hybens
    Start 6: global_enkf
1/6 Test #3: rrfs_3denvar_rdasens .............   Passed  858.84 sec
2/6 Test #6: global_enkf ......................   Passed  2735.79 sec
3/6 Test #5: hafs_3denvar_hybens ..............   Passed  3085.84 sec
4/6 Test #4: hafs_4denvar_glbens ..............   Passed  3624.06 sec
5/6 Test #1: global_4denvar ...................   Passed  3727.23 sec
6/6 Test #2: rtma .............................   Passed  3850.52 sec

100% tests passed, 0 tests failed out of 6

Total Test time (real) = 3850.55 sec
```

_Hercules_
All GSI ctests are still pending on Hercules.   The machine has 2803 jobs in the queue.  103 are running.  2700 are pending.  Not sure when the Hercules ctests will complete.

**Comment by RussTreadon-NOAA** [comment]
_Hercules_
```
Test project /work/noaa/da/rtreadon/git/gsi/pr810/build
    Start 1: global_4denvar
    Start 2: rtma
    Start 3: rrfs_3denvar_rdasens
    Start 4: hafs_4denvar_glbens
    Start 5: hafs_3denvar_hybens
    Start 6: global_enkf
1/6 Test #3: rrfs_3denvar_rdasens .............   Passed  27630.52 sec
2/6 Test #6: global_enkf ......................   Passed  29783.16 sec
3/6 Test #2: rtma .............................   Passed  31817.00 sec
4/6 Test #5: hafs_3denvar_hybens ..............   Passed  31827.25 sec
5/6 Test #1: global_4denvar ...................   Passed  31994.43 sec
6/6 Test #4: hafs_4denvar_glbens ..............***Failed  32437.47 sec

83% tests passed, 1 tests failed out of 6

Total Test time (real) = 32437.51 sec
```
The `hafs_4denvar_glbens` failure is due to
```
The runtime for hafs_4denvar_glbens_loproc_updat is 368.637720 seconds.  This has exceeded maximum allowable threshold time of 348.262471 seconds, resulting in Failure time-thresh of the regression test.

The runtime for hafs_4denvar_glbens_hiproc_updat is 264.768721 seconds.  This has exceeded maximum allowable threshold time of 253.254322 seconds, resulting in Failure of timethresh2 the regression test.
```
This is not a fatal failure.  The `/work/noaa/stmp` fileset is know to impact executable wall time.

**Comment by RussTreadon-NOAA** [comment]
**Additional tests on Dogwood**

Run the following cases:

1. C24L31 with C12L31 ensemble
2. C96L127 
3. C192L127
4. C384L127
5. C766L127 with C384L127 ensemble

Cases 2 through 4 were run twice:  (a) `develop gsi.x` using the binary `global_berror`, (b) `feature/netcdf_berror gsi.x` using the netcdf `global_berror`.  For each case the analysis results are identical between the executables and `global_berror` format.

The two jobs for Case 5 remain pending in the queue.

Case 1 is the ultra low-resolution case to be used in g-w CI.  The `feature/netcdf_berror gsi.x` successfully ran to completion using the L31 netcdf `global_berror` file.

**Comment by RussTreadon-NOAA** [comment]
Case 5, C768L127 with C384L127 ensemble is done.  `feature/netcdf_berror gsi.x` using the netcdf `global_berror` generates analysis increments that are bitwise identical to those generated by `develop gsi.x` using the binary `global_berror`

**Comment by RussTreadon-NOAA** [comment]
@danholdaway , @DavidNew-NOAA , and @CatherineThomas-NOAA

This PR adapts code from the JEDI GSIBEC to allow `gsi.x` to read netcdf format `global_berror` files.  EIB added the netcdf `global_berror` files to their staged GSI directory.  `gsi.x` determines the file format, netcdf or binary, upon reading local file `berror_stats` in the run directory.  

@CoryMartin-NOAA created netcdf `global_berror` files for for C24L31 and C12L31.  He also created a C24L31 deterministic background along with a 3 member C12L31 ensemble.  These L31 netcdf global_berror files are in EIB's staged directory along with netcdf `global_berror` files for other resolutions.  The updated `fix` hash in this PR brings in L31 ascii fix files.  I successfully ran 3DEnVar `gsi.x` with a C24 deterministic background and a 3-member C12 ensemble.
 
GSI code review policy requires two peer reviews and approvals before PRs can be merged into `develop`.  If any of you have time, a review would be helpful.   Thanks.

File: src/gsi/m_berror_stats.f90
Comment (danholdaway) [inline src/gsi/m_berror_stats.f90:14]: Why is this change from 2014 appearing?

Comment (RussTreadon-NOAA) [reply src/gsi/m_berror_stats.f90:14]: @CoryMartin-NOAA asked a[ similar question](https://github.com/NOAA-EMC/GSI/pull/810#discussion_r1866029539).  

Routine `setcorchem_` is in both `develop` ([here](https://github.com/NOAA-EMC/GSI/blob/9ada88e8b080c37a5146092e6a0a8dfc45b6654b/src/gsi/m_berror_stats.f90#L783)) and this PR ([here](https://github.com/RussTreadon-NOAA/GSI/blob/0e07d71fdd846c5590e44285f14ad4b9f23f94c9/src/gsi/m_berror_stats.f90#L1002)).  The comment above is only in this PR.  My guess is that the GSI code which Ricardo started from differs from EMC's code.

This is not an argument to retain the chem code ... just noting that this PR did not add the code.

**Review by danholdaway — APPROVED** [review]

**Comment by RussTreadon-NOAA** [comment]
@DavidNew-NOAA or @CatherineThomas-NOAA , do you have time to review this PR?  If not, let me know and I'll reach out to other developers.

**Review by DavidNew-NOAA — APPROVED** [review]
Looks fine to me, with the caveat that I don't know my way around GSI very well.

**Comment by RussTreadon-NOAA** [comment]
Thank you @DavidNew-NOAA !

**Comment by RussTreadon-NOAA** [comment]
Install `RussTreadon-NOAA:feature/netcdf_berror` at 63df6ff4 on Dogwood, Hera, and Orion.  Use `develop` at 9ada88e8 as _contrl_.  `RussTreadon-NOAA:feature/netcdf_berror` is _updat_.  Run ctests with the following results

_Dogwood_
```
Test project /lfs/h2/emc/da/noscrub/russ.treadon/git/gsi/pr810/build
    Start 1: global_4denvar
    Start 2: rtma
    Start 3: rrfs_3denvar_rdasens
    Start 4: hafs_4denvar_glbens
    Start 5: hafs_3denvar_hybens
    Start 6: global_enkf
1/6 Test #3: rrfs_3denvar_rdasens .............   Passed  907.34 sec
2/6 Test #6: global_enkf ......................   Passed  971.99 sec
3/6 Test #4: hafs_4denvar_glbens ..............   Passed  1454.36 sec
4/6 Test #5: hafs_3denvar_hybens ..............   Passed  1454.78 sec
5/6 Test #2: rtma .............................   Passed  1929.28 sec
6/6 Test #1: global_4denvar ...................   Passed  2822.58 sec

100% tests passed, 0 tests failed out of 6

Total Test time (real) = 2822.59 sec
```

_Hera_
```
Test project /scratch1/NCEPDEV/da/Russ.Treadon/git/gsi/pr810/build
    Start 1: global_4denvar
    Start 2: rtma
    Start 3: rrfs_3denvar_rdasens
    Start 4: hafs_4denvar_glbens
    Start 5: hafs_3denvar_hybens
    Start 6: global_enkf
1/6 Test #3: rrfs_3denvar_rdasens .............   Passed  495.79 sec
2/6 Test #6: global_enkf ......................   Passed  761.26 sec
3/6 Test #2: rtma .............................   Passed  971.18 sec
4/6 Test #5: hafs_3denvar_hybens ..............   Passed  1226.19 sec
5/6 Test #4: hafs_4denvar_glbens ..............   Passed  1349.04 sec
6/6 Test #1: global_4denvar ...................   Passed  1926.24 sec

100% tests passed, 0 tests failed out of 6

Total Test time (real) = 1926.26 sec
```

_Orion_
```
Test project /work2/noaa/da/rtreadon/git/gsi/pr810/build
    Start 1: global_4denvar
    Start 2: rtma
    Start 3: rrfs_3denvar_rdasens
    Start 4: hafs_4denvar_glbens
    Start 5: hafs_3denvar_hybens
    Start 6: global_enkf
1/6 Test #6: global_enkf ......................   Passed  909.31 sec
2/6 Test #3: rrfs_3denvar_rdasens .............   Passed  1026.53 sec
3/6 Test #2: rtma .............................   Passed  1687.96 sec
4/6 Test #5: hafs_3denvar_hybens ..............   Passed  2719.90 sec
5/6 Test #4: hafs_4denvar_glbens ..............   Passed  3021.75 sec
6/6 Test #1: global_4denvar ...................   Passed  3722.15 sec

100% tests passed, 0 tests failed out of 6

Total Test time (real) = 3722.17 sec
```

In the above tests `RussTreadon-NOAA:feature/netcdf_berror gsi.x` was run using the binary `global_error` file.   Repeat the above ctests using the netcdf `global_berror` when running `RussTreadon-NOAA:feature/netcdf_berror gsi.x`.  Results are as follows

_Dogwood_
```
Test project /lfs/h2/emc/da/noscrub/russ.treadon/git/gsi/pr810/build
    Start 1: global_4denvar
    Start 2: rtma
    Start 5: hafs_3denvar_hybens
    Start 4: hafs_4denvar_glbens
    Start 6: global_enkf
    Start 3: rrfs_3denvar_rdasens
1/6 Test #3: rrfs_3denvar_rdasens .............   Passed  1026.80 sec
2/6 Test #6: global_enkf ......................   Passed  1223.14 sec
3/6 Test #5: hafs_3denvar_hybens ..............   Passed  1395.08 sec
4/6 Test #2: rtma .............................   Passed  1510.72 sec
5/6 Test #4: hafs_4denvar_glbens ..............   Passed  1634.73 sec
6/6 Test #1: global_4denvar ...................   Passed  2043.95 sec

100% tests passed, 0 tests failed out of 6

Total Test time (real) = 2044.07 sec
```

_Hera_
```
era(hfe05):/scratch1/NCEPDEV/da/Russ.Treadon/git/gsi/pr810/build$ tail -f stdout_ctest_netcdf.txt
    Start 4: hafs_4denvar_glbens
    Start 5: hafs_3denvar_hybens
    Start 2: rtma
    Start 6: global_enkf
    Start 3: rrfs_3denvar_rdasens
1/6 Test #3: rrfs_3denvar_rdasens .............   Passed  496.49 sec
2/6 Test #6: global_enkf ......................   Passed  765.35 sec
3/6 Test #2: rtma .............................   Passed  970.30 sec
4/6 Test #5: hafs_3denvar_hybens ..............   Passed  1117.35 sec
5/6 Test #4: hafs_4denvar_glbens ..............   Passed  1350.57 sec
6/6 Test #1: global_4denvar ...................   Passed  1928.94 sec

100% tests passed, 0 tests failed out of 6

Total Test time (real) = 1928.97 sec
```

_Orion_
```
Test project /work2/noaa/da/rtreadon/git/gsi/pr810/build
    Start 1: global_4denvar
    Start 4: hafs_4denvar_glbens
    Start 5: hafs_3denvar_hybens
    Start 2: rtma
    Start 3: rrfs_3denvar_rdasens
    Start 6: global_enkf
1/6 Test #6: global_enkf ......................   Passed  788.22 sec
2/6 Test #3: rrfs_3denvar_rdasens .............   Passed  966.63 sec
3/6 Test #2: rtma .............................   Passed  1748.11 sec
4/6 Test #5: hafs_3denvar_hybens ..............   Passed  2780.33 sec
5/6 Test #4: hafs_4denvar_glbens ..............   Passed  3020.22 sec
6/6 Test #1: global_4denvar ...................   Passed  3662.79 sec

100% tests passed, 0 tests failed out of 6

Total Test time (real) = 3662.81 sec
```

Additionally, the `develop` and `RussTreadon-NOAA:feature/netcdf_berror` `gsi.x` were run using input from the operational 20241203 06Z GDAS for various analysis grid resolutions of C96, C192 and C384.  The `RussTreadon-NOAA:feature/netcdf_berror gsi.x` was run twice: (1) binary `global_berror`, (2) netcdf `global_berror`.  Analysis results from the three runs at each resolution are identical.  Wall times are comparable.  The size of the stdout files are comparable.  

The operational 20241206 06Z GDAS was run at C768.  Again, identical analysis results are generated with comparable wall times and stdout sizes.

**Pull request merged**
